### PR TITLE
[Large] Class Route & Completed Member (identity) 

### PR DESCRIPTION
**Pull request by leonidbogin**
- Completed the possibility of going back
- Added the Route class to work with links
- Completed work with identity
- And small fixes

### CODE REVIEW

File: DIMS-Core/DIMS-Core/Controllers/MemberController.cs [hunk 1]
```diff
@@ -35,127 +35,161 @@ public class MemberController : Controller
...
-        [Authorize]
         public async Task<IActionResult> Index()
         {
             var searchResult = await memberService.GetAll();
             var model = mapper.Map<IEnumerable<MemberViewModel>>(searchResult);
 
+            ViewBag.Route = new Route()
+            {
+                Controller = "members"
+            };
```
Comment (maximsan) [inline DIMS-Core/DIMS-Core/Controllers/MemberController.cs:46]: Why do you need this ?

File: DIMS-Core/DIMS-Core/Controllers/MemberController.cs [hunk 2]
```diff
@@ -35,127 +35,161 @@ public class MemberController : Controller
...
         [HttpPost("create")]
         [ValidateAntiForgeryToken]
-        public async Task<IActionResult> Create([FromForm]AddMemberViewModel model)
+        public async Task<IActionResult> Create([FromForm] AddMemberViewModel model, [FromQuery] Route route)
         {
             if (!ModelState.IsValid)
             {
                 ViewBag.Directions = await directionService.GetAll();
+                ViewBag.Roles = userService.GetRoles();
+                ViewBag.Route = route;
```
Comment (maximsan) [inline DIMS-Core/DIMS-Core/Controllers/MemberController.cs:72]: Why do you repeat it? Please, follow DRY

File: DIMS-Core/DIMS-Core/Views/Member/Edit.cshtml [hunk 3]
```diff
@@ -2,16 +2,36 @@
...
+    var bottomButtons = new BottomButtonsViewModel()
+    {
+        BackAction = ViewBag.Route.BackAction,
+        BackController = ViewBag.Route.BackController ??= "members"
+    };
+    var backRouteData = new Dictionary<string, string>
+    {
+        { "BackAction", bottomButtons.BackAction },
+        { "BackController", bottomButtons.BackController }
+    };
```
Comment (maximsan) [inline DIMS-Core/DIMS-Core/Views/Member/Edit.cshtml:14]: What do you think about some helpers for these? You have basically repeatable logic in huge amount of cshtml files and it looks annoying.

File: DIMS-Core/DIMS-Core/Views/Member/Create.cshtml [hunk 4]
```diff
@@ -2,15 +2,32 @@
...
+                <label class="control-label">Role</label>
+                <select asp-for="Role" class="form-control">
+                    @foreach (var role in ViewBag.Roles)
+                    {
+                        var selected = role.Name.Equals(Model.Role) ? "selected" : null;
+                        <option value="@role.Name" selected="@selected">@role.Name</option>
+                    }
+                </select>
+                <span asp-validation-for="Role" class="text-danger"></span>
+            </div>
```
Comment (maximsan) [inline DIMS-Core/DIMS-Core/Views/Member/Create.cshtml:29]: I guess this part of code repeated several times too. Could you create a partial view for this ? What do you think ?

File: DIMS-Core/DIMS-Core/Models/Route/Route.cs [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+        [BindingBehavior(BindingBehavior.Optional)]
+        public string BackAction { get; set; }
+        [BindingBehavior(BindingBehavior.Optional)]
+        public int? TaskId { get; set; }
+        [BindingBehavior(BindingBehavior.Optional)]
+        public int? UserId { get; set; }
+        [BindingBehavior(BindingBehavior.Optional)]
+        public int? UserTaskId { get; set; }
+        [BindingBehavior(BindingBehavior.Optional)]
+        public int? TaskTrackId { get; set; }
```
Comment (maximsan) [inline DIMS-Core/DIMS-Core/Models/Route/Route.cs:26]: You have here the same props as in `BackToListViewModel` or smth like this. So I propose you to create common abstract class to share the props and to dot repeat them several times.

**Review by maximsan — CHANGES_REQUESTED** [review]
Hello! I left some notes. please looks through them. Thanks!

File: DIMS-Core/DIMS-Core/Models/Route/Route.cs [hunk 6]
```diff
@@ -0,0 +1,28 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Mvc.ModelBinding;
+
+namespace DIMS_Core.Models.Route
+{
+    public class Route
```
Comment (PushMeNow) [inline DIMS-Core/DIMS-Core/Models/Route/Route.cs:9]: This class is pretty ridiculous. You try to put all existing route parameters in one class. But it is bad practice. Just think about it. In each method you need ONLY one parameter from this class but what about other params??)) You will not use them so why is they there?)) What is the goal?

File: DIMS-Core/DIMS-Core/Views/Member/Index.cshtml [hunk 7]
```diff
@@ -61,16 +53,25 @@
...
                                 </div>
                                 <div class="col">
-                                    @Html.ActionLink("Tasks", "Tasks", new {id = item.UserId}, new {@class = "btn btn-primary w-100 m-2"})
-                                    @Html.ActionLink("Delete", "Delete", new {id = item.UserId}, new {@class = "btn btn-danger w-100 m-2"})
+                                    @Html.ActionLink("Progress", "Progress", "Member", new Route() { UserId = item.UserId }, new { @class = "btn btn-primary w-100 m-2" })
                                 </div>
                             </div>
-                        </td>
-                    }
+                            @if (User.IsInRole("admin"))
```
Comment (PushMeNow) [inline DIMS-Core/DIMS-Core/Views/Member/Index.cshtml:62]: Write please extension to check role for user like: IsInRole("admin") -> IsAdmin();

**Review by PushMeNow — CHANGES_REQUESTED** [review]
Commit is realy strange. You created much logic but I didn't unnderstand the reason...
I did some comments fix them. Will see after that.

**Pull request closed without merging**